nrf54h20: add exmif peripheral DTS description and sample configuration

e-rk opened the pull request.

e-rk committed:
samples: drivers: jesd216: add nRF54H20 overlay

Configure the sample for nRF54H20 CPUAPP.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
dts: nordic: add EXMIF peripheral description to nRF54H20

Added EXMIF peripheral DTS description and bindings.
The peripheral operates as an SPI device.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

gmarull commented on samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -0,0 +1,25 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&uart136 {
+	status = "okay";
+};
```
why is this needed?

e-rk replied:
I think it is a leftover. I see that `uart136` is enabled by default now.

e-rk replied:
Removed

gmarull commented on samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -0,0 +1,25 @@
+	status = "okay";
+};
+
+&gpio6 {
+	status = "okay";
+};
```
same: where is this used in this overlay?

gmarull commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:
```diff
@@ -195,3 +195,27 @@ ipc0: &cpuapp_cpurad_ipc {
+	pinctrl-names = "default";
+	mx25uw63: mx25uw6345g@0 {
+		compatible = "jedec,spi-nor";
+		status = "disabled";
```
why is it disabled at board level?

e-rk replied:
I can enable it by default, if that's the recommendation. On nRF54H20 DK the pins used by EXMIF/external flash are not connected to anything else.

gmarull replied:
at board level things are usually enabled by default.

e-rk replied:
Enabled EXMIF by default, but flash node is kept disabled, like it was done here: https://github.com/zephyrproject-rtos/zephyr/blame/main/boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts#L135

gmarull replied:
why?

e-rk replied:
I wanted to avoid having `CONFIG_SPI_NOR` enabled in all samples by default. Setting status okay would result in `CONFIG_SPI_NOR` being enabled by default when `CONFIG_FLASH=y`. On current nRF54H20 DKs, powering up the flash chip requires some additional steps from the user.

gmarull commented on include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:
```diff
@@ -126,6 +126,28 @@
 #define NRF_FUN_QSPI_IO2 33U
 /** QSPI IO3 */
 #define NRF_FUN_QSPI_IO3 34U
+/** EXMIF CK */
+#define NRF_FUN_EXMIF_CK 35U
+/** EXMIF DQ0 */
+#define NRF_FUN_EXMIF_DQ0 36U
+/** EXMIF DQ1 */
+#define NRF_FUN_EXMIF_DQ1 37U
+/** EXMIF DQ2 */
+#define NRF_FUN_EXMIF_DQ2 38U
+/** EXMIF DQ3 */
+#define NRF_FUN_EXMIF_DQ3 39U
+/** EXMIF DQ4 */
+#define NRF_FUN_EXMIF_DQ4 40U
+/** EXMIF DQ5 */
+#define NRF_FUN_EXMIF_DQ5 41U
+/** EXMIF DQ6 */
+#define NRF_FUN_EXMIF_DQ6 42U
+/** EXMIF DQ7 */
+#define NRF_FUN_EXMIF_DQ7 43U
+/** EXMIF CS0 */
+#define NRF_FUN_EXMIF_CS0 44U
+/** EXMIF CS1 */
+#define NRF_FUN_EXMIF_CS1 45U
```
content in the wrong commit?

e-rk replied:
This is the first commit that makes use of those definitions, so I put them here. Should those definitions be in a separate commit?

gmarull replied:
How does the pinctrl driver configure exmif?

e-rk replied:
On nRF54H20 it's not needed. The configuration is done by nrf-regtool.

anangl replied:
I think it would be better to add this definitions in a separate commit, with a note in its message explaining why no changes in `pinctrl_nrf.c` are needed.

gmarull replied:
> I think it would be better to add this definitions in a separate commit, with a note in its message explaining why no changes in `pinctrl_nrf.c` are needed.

agreed

e-rk replied:
Split the change to a separate commit.

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -987,6 +987,19 @@
 				interrupts = <470 NRF_DEFAULT_IRQ_PRIORITY>;
 			};
+
+			exmif: spi@95000 {
```
Please keep the nodes ordered by address and locate this node right after the `usbhs` one.

e-rk replied:
Changed the order

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -987,6 +987,19 @@
 			};
+
+			exmif: spi@95000 {
+				compatible = "nordic,nrf-exmif", "snps,designware-spi";
```
I don't think that "snps,designware-spi" should be listed here. It is not actually possible to use this peripheral that way without Nordic specific stuff.

e-rk replied:
Removing the compatible results in spi_dw.c not being compiled and `DT_INST_FOREACH_STATUS_OKAY` not picking up the node.

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -987,6 +987,19 @@
+				reg = <0x95000 0x500 0x95500 0xb00>;
+				reg-names = "wrapper", "core";
+				interrupts = <149 NRF_DEFAULT_IRQ_PRIORITY>;
+				clock-frequency = <DT_FREQ_M(400)>;
```
what is the clock source here? looks like this should be referencing a clock?

e-rk replied:
The clock source frequency is fixed and the application domain will only be able to turn it on or off. Currently there is no service that allows turning it on or off, but it is enabled by default.

gmarull replied:
I mean that from a DT perspective, looks like there should be a clock that runs at 400MHz defined somewhere, unless this peripheral has a contained 400MHz clock

gmarull replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/cdc0dbfb-fddf-443e-bec2-dcc1fd55df94)

gmarull replied:
so it looks like you should define a `fixed-clock` from an application perspective (similar to fll16m), as it's manages by sysctrl, then reference it as `clocks = <&hsfll_exmif>;`

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/79cf411a-e6f7-4ecc-8d1c-a942ee076786)

e-rk replied:
Added the clock source.

e-rk committed:
soc: nrf54h: Enable SPI DW HSSI register layout

The nRF54H20 implements a variant of the SPI DW peripheral that
has slightly different register layout. Enable it in the defconfig.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
boards: nordic: add mx25uw6345g flash node to nRF54H20dk

The nRF54H20dk has mx25uw6345g flash device on board. Added DTS
description and pinctrl configuration.

The flash device node is disabled by default. The flash device needs to
be powered-on using the nRF Connect for Desktop Board Configurator
application.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
include: dts: nrf: add EXMIF pinctrl definitions

Added EXMIP pinctrl definitions, which allow selecting pin functions by
name in DTS. The definitions are added, but not used in pinctrl_nrf.c.

The nrf-regtool reads the pinctrl configuration and applies the settings
using different mechanisms.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk commented:
I added one more commit which modified the nRF54H20 Kconfig.defconfig. It is needed to select the correct register layout for the SPI DW peripheral.

gmarull commented on boards/nordic/nrf54h20dk/Kconfig.defconfig:
```diff
@@ -15,3 +15,6 @@ config BT_CTLR
 
 endif # BOARD_NRF54H20DK_NRF54H20_CPURAD
+
+config SPI_DW_HSSI
+	default y if SPI_DW
```
is this really a board-level defconfig?

e-rk replied:
Right, that was the wrong one. Moved to soc defconfig.

gmarull commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -93,6 +93,12 @@
 			clock-frequency = <DT_FREQ_M(16)>;
 		};
+
+		exmif_hsfll: exmif_hsfll {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <DT_FREQ_M(400)>;
+		};
```
note: commit message says app needs to turn this on/off, so this is not really a fixed-clock, it will require a clock control driver that uses nrfs underneath. Looks like we'll need a new compatible, eg `nordic,nrf-hsfll-remote` or similar and a driver so we can use `clock_control_on/off`. @anangl any comments here?

e-rk replied:
Should this be a part of this PR, or could this be addressed later?

anangl replied:
AFAICT the driver only supports the fixed clock frequency now: https://github.com/zephyrproject-rtos/zephyr/blob/0b87cb1545d5ec99b4f5e8368bb04dfbee4098b1/drivers/spi/spi_dw.c#L636-L641 so I think there is no point in trying to provide proper handling of that HSFLL here. I think the initial approach with the `clock-frequency` property in the `exmif` node was fine. Especially taking into account that the main idea behind this PR is to just take advantage of the already existing `spi_dw` driver to use the EXMIF peripheral in its basic SPI mode, temporarily, until we have a proper EXMIF driver.

e-rk replied:
Looking at the above snippet the previous approach with `clock-frequency` gets the job done just the same as `clocks`. I'll stay with the current approach (`clocks`) unless some problem emerges.

anangl replied:
Yes, it does the job in the `spi_dw` driver context, but the HSFLL is not actually a fixed clock, as its output frequency can be changed, so the added node is a bit misleading.

e-rk replied:
Okay, I restored it to the previous approach.

anangl commented on samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&mx25uw63 {
```
The order of commits is wrong. The above node label is added in the third commit but it is referred in the first one.

anangl replied:
nit: also "soc: nrf54h: Enable SPI DW HSSI register layout" should be before "boards: nordic: add mx25uw6345g flash node to nRF54H20dk" that activates the spi_dw driver.

e-rk replied:
Done
